File: Python/No20.py
class Solution:
    def isValid(self, s: str) -> bool:
        if not s: return True
        if len(s) % 2 == 1: return False
        stack = []
        table = {')': '(', '}': '{', ']': '['}
        for c in s:
            if c in ['(', '{', '[']:
                stack.append(c)
            elif c in [')', '}', ']']:
                if len(stack) != 0:
                    top = stack.pop()
                    if table[c] != top:
                        return False
                else:
                    return False
        if len(stack) != 0:
            return False
        return True


s = Solution()
print(s.isValid("()[{}]"))

File: Java/src/No20.java
/*
 * Copyright (c) 2021, HuXiaozhong. All rights reserved.
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 */

import java.util.Stack;

public class No20 {
    public boolean isValid(String s) {
        Stack<Character> stack = new Stack<>();
        for (char c : s.toCharArray()) {
            if (c == '{' || c == '(' || c == '[') stack.add(c);
            else {
                if (stack.isEmpty()) return false;
                char top = stack.pop();
                if (c == '}' && top != '{') return false;
                else if (c == ')' && top != '(') return false;
                else if (c == ']' && top != '[') return false;
            }
        }
        return stack.isEmpty();
    }
}
